70_FLIPFluids/main.cpp

```diff
@@ -0,0 +1,2240 @@
+			m_logger->log("Shader %s not found!", ILogger::ELL_ERROR, filePath);
+			exit(-1);
+		}
+		
+		const auto assets = bundle.getContents();
+		assert(assets.size() == 1);
+		smart_refctd_ptr<ICPUShader> shaderSrc = IAsset::castDown<ICPUShader>(assets[0]);
+
+		smart_refctd_ptr<ICPUShader> shader = shaderSrc;
+		if (entryPoint != "main")
+		{
+			auto compiler = make_smart_refctd_ptr<asset::CHLSLCompiler>(smart_refctd_ptr(m_system));
+			CHLSLCompiler::SOptions options = {};
+			options.stage = shaderSrc->getStage();
+			if (!(options.stage == IShader::E_SHADER_STAGE::ESS_COMPUTE || options.stage == IShader::E_SHADER_STAGE::ESS_FRAGMENT))
+				options.stage = IShader::E_SHADER_STAGE::ESS_VERTEX;
+			options.targetSpirvVersion = m_device->getPhysicalDevice()->getLimits().spirvVersion;
+			options.spirvOptimizer = nullptr;
+			options.debugInfoFlags |= IShaderCompiler::E_DEBUG_INFO_FLAGS::EDIF_SOURCE_BIT;
+			options.preprocessorOptions.sourceIdentifier = shaderSrc->getFilepathHint();
+			options.preprocessorOptions.logger = m_logger.get();
+			options.preprocessorOptions.includeFinder = compiler->getDefaultIncludeFinder();
+
+			std::string dxcOptionStr[] = {"-E " + entryPoint};
+			options.dxcOptions = std::span(dxcOptionStr);
+
+			shader = compiler->compileToSPIRV((const char*)shaderSrc->getContent()->getPointer(), options);
+		}
+
+		return m_device->createShader(shader.get());
+	}
+
+	bool createBuffer(smart_refctd_ptr<IGPUBuffer>& buffer, video::IGPUBuffer::SCreationParams& params)
+	{
+		buffer = m_device->createBuffer(std::move(params));
+		if (!buffer)
+			return logFail("Failed to create GPU buffer of size %d!\n", params.size);
+
+		video::IDeviceMemoryBacked::SDeviceMemoryRequirements reqs = buffer->getMemoryReqs();
+		reqs.memoryTypeBits &= m_physicalDevice->getDeviceLocalMemoryTypeBits();
+
+		auto bufMem = m_device->allocate(reqs, buffer.get());
+		if (!bufMem.isValid())
+			return logFail("Failed to allocate device memory compatible with gpu buffer!\n");
+
+		return true;
+	}
+
+	bool initGraphicsPipeline()
+	{
+		m_renderSemaphore = m_device->createSemaphore(m_submitIx);
+		if (!m_renderSemaphore)
+			return logFail("Failed to create render semaphore!\n");
+			
+		ISwapchain::SCreationParams swapchainParams{
+			.surface = m_surface->getSurface()
+		};
+		if (!swapchainParams.deduceFormat(m_physicalDevice))
+			return logFail("Could not choose a surface format for the swapchain!\n");
+
+		const static IGPURenderpass::SCreationParams::SSubpassDependency dependencies[] = {
+			{
+				.srcSubpass = IGPURenderpass::SCreationParams::SSubpassDependency::External,
+				.dstSubpass = 0,
+				.memoryBarrier = {
+					.srcStageMask = PIPELINE_STAGE_FLAGS::LATE_FRAGMENT_TESTS_BIT,
+					.srcAccessMask = ACCESS_FLAGS::DEPTH_STENCIL_ATTACHMENT_WRITE_BIT,
+					.dstStageMask = PIPELINE_STAGE_FLAGS::EARLY_FRAGMENT_TESTS_BIT,
+					.dstAccessMask = ACCESS_FLAGS::DEPTH_STENCIL_ATTACHMENT_WRITE_BIT | ACCESS_FLAGS::DEPTH_STENCIL_ATTACHMENT_READ_BIT
```

Comment:
this is not enough, vertex_shader needs to also be in the source stage if you do programmable pulling with BDA (otherwise it would need to be the vertex_input if you pulled particle data with instance divisor of 1)

destination access also need the velocity grid image sampling and SHADER_READ (for BDA, if you were using vertex attributes then the vertex attribute input mask)

Reply:
but thankfully you never used vertex attribute inputs and started off with programmable pulling right away